go/consensus/tendermint/apps/staking/messages.go

```diff
@@ -41,6 +42,54 @@ func (app *stakingApplication) changeParameters(ctx *api.Context, msg interface{
 		return nil, fmt.Errorf("staking: failed to validate consensus parameters: %w", err)
 	}
 
+	// Do any necessary state migrations.
+	if changes.MinCommissionRate != nil {
+		var epoch beacon.EpochTime
+		epoch, err = ctx.AppState().GetCurrentEpoch(ctx)
+		if err != nil {
+			return nil, fmt.Errorf("staking: failed to load epoch")
+		}
+		// On MinCommissionRate update, the staking state needs to be updated to ensure all
+		// commission rates and bounds are above the new min commission rate.
+		var addresses []staking.Address
+		addresses, err = state.Addresses(ctx)
```

Comment:
One way to make this less expensive would be to store `CommissionSchedule`s separately, under its own storage key and provide a staking state API method to return all commission schedules. 

To minimize the amount of breaking changes, the `SetAccount`/`(Get)Account` methods could be kept as are, and internally handle the storing/fetching of commission schedules.

Reply:
As a first step I would try to see how expensive (time and memory use) this migration step would be for Mainnet (~300k accounts) and up to 1M accounts. Storing the commission schedule separately would make all the usual (and more frequent) cases more expensive though (as additional fetches / writes would be required).

Reply:
Actually, these numbers were wrong, as the tests were done at InitChain, when the accounts were just inserted and therefore in memory already. 

It takes ~3 seconds to load the addresses, and then another ~30 seconds to just traverse all the accounts (without any inserts). This seems quite slow/expensive.

---

> Storing the commission schedule separately would make all the usual (and more frequent) cases more expensive though (as additional fetches / writes would be required). 

Could also consider a bit bigger braking change and not only store the `CommissionSchedule` separately, but also removing it from the `Escrow` account structure. 
I would assume (need to confirm if this is the case) that in most cases where accounts are queried/modified, the commission schedule is not relevant. So the additional fetches/writes would only happen in places that actually deal with both accounts and commission schedules. This would however probably be a quite big breaking change, so I'm not sure if it's even to be considered.

---

Do you have any other ideas thoughts?

Reply:
One possible (non breaking) option would be adding an index of accounts with commission schedule configured to the storage. That way, only the accounts with commission schedule set would need to be queried (only ~170 such accounts on the mainnet). Although I think that that nothing prevents accounts from mass configuring the commission schedules, so not guaranteed to be effective in future as well.

Reply:
> One possible (non breaking) option would be adding an index of accounts with commission schedule configured to the storage.

I think this would be ok for now. We could also consider making it more expensive (but not disruptive) to set the commission schedule, e.g. maybe require that the account has at least the entity threshold staked?

Reply:
Agreed. The stake requirement can probably be entity threshold plus validator node threshold.

Reply:
> One possible (non breaking) option would be adding an index of accounts with commission schedule configured to the storage.

Added.

> We could also consider making it more expensive (but not disruptive) to set the commission schedule, e.g. maybe require that the account has at least the entity threshold staked

Added.

---

Will also re-do the benchmarks a bit later today.

>

Reply:
By using the new method on recent mainnet data, only `167` accounts need to be queried and updated. It takes tens of milliseconds to complete it.